ssl/event_queue.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include "internal/event_queue.h"
+#include "crypto/sparse_array.h"
+#include "ssl_local.h"
+
+struct ossl_event_subscription_s {
+    struct ossl_event_subscription_s *prev, *next;
+    ossl_event_callback_fn *cb;
+    void *cbarg;
+    OSSL_EVENT_QUEUE *queue;
+    uint32_t type;
+};
+
+DEFINE_SPARSE_ARRAY_OF(OSSL_EVENT_SUBSCRIPTION);
+
+struct ossl_event_queue_s {
+    PRIORITY_QUEUE_OF(OSSL_EVENT) *events;
+    SPARSE_ARRAY_OF(OSSL_EVENT_SUBSCRIPTION) *subscribers;
+};
+
+static int event_compare(const OSSL_EVENT *a, const OSSL_EVENT *b)
+{
+    int i = ossl_time_compare(a->when, b->when);
+
+    if (i != 0)
+        return i;
+    if (a->priority > b->priority)
+        return 1;
+    if (a->priority < b->priority)
+        return -1;
+    return 0;
+}
+
+OSSL_EVENT_QUEUE *ossl_event_queue_new(void)
+{
+    OSSL_EVENT_QUEUE *r = OPENSSL_malloc(sizeof(*r));
+
+    if (r != NULL) {
+        r->events = ossl_pqueue_OSSL_EVENT_new(&event_compare);
+        r->subscribers = ossl_sa_OSSL_EVENT_SUBSCRIPTION_new();
+        if (r->events == NULL || r->subscribers == NULL) {
+            ossl_event_queue_free(r);
+            return NULL;
+        }
+    }
+    return r;
+}
+
+void ossl_event_queue_free(OSSL_EVENT_QUEUE *queue)
+{
+    OSSL_EVENT *e;
+
+    /* This isn't remotely correct yet */
+    if (queue != NULL) {
+        if (queue->events != NULL) {
+            while ((e = ossl_pqueue_OSSL_EVENT_pop(queue->events)) != NULL)
+                OPENSSL_free(e);
+            ossl_pqueue_OSSL_EVENT_free(queue->events);
+        }
+        ossl_sa_OSSL_EVENT_SUBSCRIPTION_free(queue->subscribers);
+        OPENSSL_free(queue);
+    }
+}
+
+OSSL_EVENT *ossl_event_new(uint32_t type, uint32_t priority,
+                           OSSL_TIME when, void *ctx,
+                           void *payload, size_t payload_size)
+{
+    OSSL_EVENT *e = OPENSSL_malloc(sizeof(*e));
+
+    if (e != NULL) {
+        ossl_event_set(e, type, priority, when, ctx, payload, payload_size);
+        e->flag_dynamic = 1;
+    }
+    return e;
+}
+
+void ossl_event_free(OSSL_EVENT *event)
+{
+    if (event != NULL && event->flag_dynamic)
+        OPENSSL_free(event);
+}
+
+int ossl_event_queue_add(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event){
+    return ossl_pqueue_OSSL_EVENT_push(queue->events, event, &event->ref);
```

Comment:
So in this event queue design adding events to the queue does not also dispatch them. For timed events that haven't "fired" yet this is fine. What about "immediate" events? How do we handle those? Presumably they are represented internally with a "when" of "now" (or is that necessary....do we continually want to be querying the system time for this type of event?)

Reply:
I'd represent them with a "when" of zero rather than now (which should be wrapped in a macro).

I see running them immediately as a trivial change, if we wanted to.  There are a few potential concerns with doing so:
1. Raising two events of different priority simultaneously could go awry if done in reverse order (more likely if different pieces of code raise them -- e.g. during an event call back sequence).
2. It removes control of the dispatching from the calling application (i.e. stuff happens without calling _tick).
3. It could introduce recursion (i.e. locking) problems (e.g. calling an application callback which raises an immediate event via some action it does).

I've also figured out a way to avoid them hitting the main event queue if the performance loss of doing so is discovered to be a problem